### ISSUE TITLE
Migration requirements make it hard to rollback Horizon

### ISSUE THREAD

**Issue by bartekn** [stellar/go#909]
Changes introduced in https://github.com/stellar/go/pull/808 make it hard to rollback Horizon in case of errors. I'm wondering if we should remove checking for extra migrations as they always be backward compatible with the previous version.

CC: @ire-and-curses @tomquisel

**Comment by ire-and-curses** [comment]
Would you add some details about the situation that makes it hard to rollback?

**Comment by bartekn** [comment]
Sure. I deployed Horizon 0.17.0 release candidate to horizon-testnet.stellar.org. Migrations were performed during deploy. Then I noticed #908 and decided to rollback to 0.16.0. It didn't start because it saw extra migrations (from 0.17.0). I had to manually remove a row in `gorp_migrations` to make it start. I couldn't just run `db migrate down 1` because 0.16.0 has no knowledge about 0.17.0 migrations.

**Comment by ire-and-curses** [comment]
RIght. Yeah, the expectation was that you would run `db migrate down ` using a version of the code equal to or later to the DB. So I think if you had done this before rolling back to 0.16.0 it would have worked.

I can see how this is an inconvenience though. If you're sure the DB will always be backwards compatible, then I agree we could remove the check. I guess I'm worried that after running the older version of the code, the new columns will be in an inconsistent state if you moved back to the more recent version, unless you removed and re-applied the migration.

Alternatively, I could add a `force` option to allow you to run an older version of the code with a newer DB. In this case the user is taking responsibility for the cosequences.

**Comment by bartekn** [comment]
Another option would be to add migration file to `gorp_migrations` table. Then if you want to `migrate down` using older version Horizon knows how to do it.

Let's move it to backlog for now.

**Comment by bartekn** [comment]
It happened to me again when reverting 0.19.0rc in testnet. We should prioritize this, @ire-and-curses @tomquisel.

**Comment by tomquisel** [comment]
I support fixing this however you think is best (`force` option, full disable, or add migration file to that table).

**Comment by bartekn** [comment]
I'm not 100% sure yet (let's discuss this) but it seems the best option is to save migrations SQL in `gorp_migrations` table. Then we can create a new command like `horizon db migrate sync` that will sync a DB to the latest migration in a current version (migrating down using the SQL code if necessary). This will automatically revert schema to a good one when there's a need for a rollback.

**Comment by ire-and-curses** [comment]
We discussed this again internally.

- Now that our deploy has changed to a two cluster blue/green approach, this scenario isn't likely to be common
- Also noting that there is a correct workflow (migrate down before reverting to the earlier binary), it just isn't intuitive
- Solutions to store all relevant info in e.g. `gorp_migrations` might involve hacking a dependency, which would be a bit :hankey: 

@brahman81 do you still feel this is an issue to prioritise?

**Comment by ire-and-curses** [comment]
Closing, feel free to re-open if you feel strongly.
